dot/state/block_finalisation.go

```diff
@@ -0,0 +1,112 @@
+package state
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+// finalizedHashKey = FinalizedBlockHashKey + round + setID (LE encoded)
+func finalizedHashKey(round, setID uint64) []byte {
+	return append(common.FinalizedBlockHashKey, roundSetIDKey(round, setID)...)
+}
+
+// HasFinalizedBlock returns true if there is a finalised block for a given round and setID, false otherwise
+func (bs *BlockState) HasFinalizedBlock(round, setID uint64) (bool, error) {
+	return bs.db.Has(finalizedHashKey(round, setID))
+}
+
+// NumberIsFinalised checks if a block number is finalised or not
+func (bs *BlockState) NumberIsFinalised(num *big.Int) (bool, error) {
+	header, err := bs.GetFinalizedHeader(0, 0)
+	if err != nil {
+		return false, err
+	}
+
+	return num.Cmp(header.Number) <= 0, nil
+}
+
+// GetFinalizedHeader returns the finalised block header by round and setID
+func (bs *BlockState) GetFinalizedHeader(round, setID uint64) (*types.Header, error) {
+	h, err := bs.GetFinalizedHash(round, setID)
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := bs.GetHeader(h)
+	if err != nil {
+		return nil, err
+	}
+
+	return header, nil
+}
+
+// GetFinalizedHash gets the finalised block header by round and setID
+func (bs *BlockState) GetFinalizedHash(round, setID uint64) (common.Hash, error) {
+	h, err := bs.db.Get(finalizedHashKey(round, setID))
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	return common.NewHash(h), nil
+}
+
+// SetFinalizedHash sets the latest finalised block header
+// Note that using round=0 and setID=0 would refer to the latest finalised hash
+func (bs *BlockState) SetFinalizedHash(hash common.Hash, round, setID uint64) error {
+	bs.Lock()
+	defer bs.Unlock()
+
+	has, _ := bs.HasHeader(hash)
+	if !has {
+		return fmt.Errorf("cannot finalise unknown block %s", hash)
+	}
+
+	// if nothing was previously finalised, set the first slot of the network to the
+	// slot number of block 1, which is now being set as final
+	if bs.lastFinalised.Equal(bs.genesisHash) && !hash.Equal(bs.genesisHash) {
+		err := bs.setFirstSlotOnFinalisation()
+		if err != nil {
+			return err
+		}
+	}
+
+	if round > 0 {
+		go bs.notifyFinalized(hash, round, setID)
+	}
+
+	pruned := bs.bt.Prune(hash)
+	for _, rem := range pruned {
+		header, err := bs.GetHeader(rem)
+		if err != nil {
+			return err
+		}
+
+		err = bs.DeleteBlock(rem)
+		if err != nil {
+			return err
+		}
+
+		logger.Trace("pruned block", "hash", rem, "number", header.Number)
+		bs.pruneKeyCh <- header
```

Comment:
this could block if `storageState` is lagging behind or not running at all.  Do we need a timeout just in case?

Reply:
the storage state just deletes a key from the map on prune, so generally I don't think it would lag. how would you implement a timeout for this?

Reply:
```
select {
   case <-bs.pruneKeyCh:
   case <-time.After(5 * time.Second):
}
```
May be a bit overkill for this case, but that's how I would do it.

Reply:
the code is writing to the channel, and your code is reading from it? is there a way to do a timeout when writing? either way I don't think it's too necessary for this case